### PR TITLE
api: Introducing a new diagnostics API command for system VMs for CloudStack admins

### PR DESCRIPTION
**Pull request by dhlaluku**
This feature introduces a new API command that will improve troubleshooting of network issues on CloudStack hosted networks by executing network-utility commands (ping, traceroute, arping) remotely on system VMs. 

## Description

For troubleshooting purposes, CloudStack administrators may wish to execute network utility commands remotely on system VMs, or request system VMs to ping/traceroute/arping to specific addresses over specific interfaces. An API command to provide such functionalities is being developed without altering any existing APIs. The targeted system VMs for this feature are the Virtual Router (VR), Secondary Storage VM (SSVM) and the Console Proxy VM (CPVM).

FS: https://cwiki.apache.org/confluence/display/CLOUDSTACK/CloudStack+Remote+Diagnostics+API
ML discussion: https://markmail.org/message/xt7owmb2c6iw7tva

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
image

## How Has This Been Tested?
Log on to the CloudStack management server as root admin and start cloudmonkey. Sync APIs and then execute the run diagnostics command as follows:
**run diganostics ipaddress=www.shapeblue.com type=arping targetid=uuid params="-I eth0 -c 4"**

Where;
- **ipaddress** is the destination IP/Domain address to test connection to;
- **type** is the diagnostics command type to execute from remote target
- **targetid** is the uuid of the system VM from which to test
- **params** are optional command line arguments that apply to each diagnostics command type

Dev environment components:
- Platform: ACS-4.12
- Management OS: Ubuntu 18.04 LTS
- Hypervisor: 1 Host running KVM with CentOS 7

This command does not run with the CloudStack simulator and has only been tested on real hardware environments. 

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [x] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: test/integration/smoke/test_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+from marvin.lib.common import (get_domain,
+                               get_zone,
+                               get_test_template,
+                               list_ssvms,
+                               list_routers)
+
+from nose.plugins.attrib import attr
+
+
+class TestRemoteDiagnostics(cloudstackTestCase):
```
Comment (rhtyd) [inline test/integration/smoke/test_diagnostics.py]: @dhlaluku can you add this to .travis.yml, fix it to make it work/skip against simulator?

**Comment by rhtyd** [comment]
Please hold merging, while everything is LGTM -- @dhlaluku will add support for simulator with his work to make the business logic test-able and add his new marvin test with .travis.yml

**Comment by dhlaluku** [comment]
@rhtyd I have included "test_diagnostics" in the Travis file. Updated the Marvin tests to skip the negative test cases for the Simulator hypervisor.

Also defined some mocks for the positive cases, example below with Cloudmonkey;

(local) 🐵 > run diagnostics targetid=0c04cf04-8222-47fd-a883-6cf335646d64 ipaddress=8.8.8.8 type=traceroute
{
  "diagnostics": {
    "exitcode": "0",
    "stderr": "",
    "stdout": "TRACEROUTE 8.8.8.8 executed in v-1-VM"
  }
}

**Comment by rhtyd** [comment]
Fantastic @dhlaluku - let's wait for travis to complete/pass.

File: .travis.yml [hunk 2]
```diff
@@ -34,6 +34,8 @@ env:
   matrix:
     # Keep the TESTS sorted by name and grouped by type
     - TESTS="smoke/test_certauthority_root"
+
+    - TESTS="smoke/test_diagnostics"
```
Comment (rhtyd) [inline .travis.yml]: Please keep the name sorted, include your file under an existing test section/block, don't add a new job/runner.

File: core/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsAnswer.java [hunk 3]
```diff
@@ -41,12 +41,14 @@ public DiagnosticsAnswer(DiagnosticsCommand cmd, boolean result, String details)
...
                 executionDetailsMap.put(ApiConstants.STDERR, parseDetails[1].trim());
                 executionDetailsMap.put(ApiConstants.EXITCODE, String.valueOf(parseDetails[2]).trim());
-                return executionDetailsMap;
             } else {
                 throw new CloudRuntimeException("Unsupported diagnostics command type supplied");
             }
         } else {
-            throw new CloudRuntimeException("Command execution failed: " + details);
+            executionDetailsMap.put(ApiConstants.STDOUT, "");
+            executionDetailsMap.put(ApiConstants.STDERR, details );
```
Comment (rhtyd) [inline core/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsAnswer.java]: Remove space after `details`.

Comment (dhlaluku) [reply core/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsAnswer.java]: Removed

File: .travis.yml (hunk 2 above)
Comment (dhlaluku) [reply .travis.yml]: Included test to the section with routers and ssvm tests.

File: test/integration/smoke/test_diagnostics.py (hunk 1 above)
Comment (dhlaluku) [reply test/integration/smoke/test_diagnostics.py]: Tests has been added to Travis file in the section that contains tests for routers and system VMs

File: .travis.yml (hunk 2 above)
Comment (rhtyd) [reply .travis.yml]: No, you should keep it sorted by name as the file suggests.

Comment (dhlaluku) [reply .travis.yml]: - TESTS="smoke/test_accounts
             smoke/test_affinity_groups
             smoke/test_affinity_groups_projects
             smoke/test_deploy_vgpu_enabled_vm
             smoke/test_deploy_vm_iso
             smoke/test_deploy_vm_root_resize
             smoke/test_deploy_vm_with_userdata
             smoke/test_deploy_vms_with_varied_deploymentplanners
             **smoke/test_diagnostics**
             smoke/test_disk_offerings
             smoke/test_dynamicroles
             smoke/test_global_settings
             smoke/test_guest_vlan_range"

**Comment by dhlaluku** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

File: .travis.yml [hunk 4]
```diff
@@ -34,7 +34,7 @@ env:
   matrix:
     # Keep the TESTS sorted by name and grouped by type
     - TESTS="smoke/test_certauthority_root"
-    
+
```
Comment (rhtyd) [inline .travis.yml]: Remove this newline

**Comment by dhlaluku** [comment]
@blueorangutan test

**Pull request merged**